src/module.c

```diff
@@ -2436,12 +2441,21 @@ static void moduleCloseKey(RedisModuleKey *key) {
     int signal = SHOULD_SIGNAL_MODIFIED_KEYS(key->ctx);
     if ((key->mode & REDISMODULE_WRITE) && signal)
         signalModifiedKey(key->ctx->client,key->db,key->key);
-    if (key->iter) zfree(key->iter);
-    RM_ZsetRangeStop(key);
-    if (key && key->value && key->value->type == OBJ_STREAM &&
-        key->u.stream.signalready) {
-        /* One of more RM_StreamAdd() have been done. */
-        signalKeyAsReady(key->db, key->key, OBJ_STREAM);
+    if (key->value) {
+        switch (key->value->type) {
+        case OBJ_LIST:
+            if (key->iter) listTypeReleaseIterator(key->iter);
+            break;
+        case OBJ_ZSET:
+            RM_ZsetRangeStop(key);
+            break;
+        case OBJ_STREAM:
+            if (key->iter) zfree(key->iter);
+            if (key->u.stream.signalready)
+                /* One of more RM_StreamAdd() have been done. */
+                signalKeyAsReady(key->db, key->key, OBJ_STREAM);
+            break;
```

Comment:
please make sure it's not possible for key->value to be already released, and key->iter still be allocated

Reply:
Adding an assert.